server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestClusterStateAction.java

```diff
@@ -107,20 +114,36 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
-                }
-                builder.field(Fields.CLUSTER_NAME, response.getClusterName().value());
-                ToXContent.Params params = new ToXContent.DelegatingMapParams(
-                    singletonMap(Metadata.CONTEXT_MODE_PARAM, Metadata.CONTEXT_MODE_API), request);
-                response.getState().toXContent(builder, params);
-                builder.endObject();
-                return new BytesRestResponse(RestStatus.OK, builder);
-            }
+        return channel -> client.admin().cluster().state(clusterStateRequest, new RestActionListener<>(channel) {
+
+                    @Override
+                    protected void processResponse(ClusterStateResponse response) {
+                        final long startTimeMs = threadPool.relativeTimeInMillis();
+                        // Process serialization on MANAGEMENT pool since the serialization of the cluster state to XContent
+                        // can be too slow to execute on an IO thread
+                        threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(
+                                ActionRunnable.wrap(this, l -> new RestBuilderListener<ClusterStateResponse>(channel) {
+                                    @Override
+                                    public RestResponse buildResponse(final ClusterStateResponse response,
+                                                                      final XContentBuilder builder) throws Exception {
+                                        if (clusterStateRequest.local() == false &&
+                                                threadPool.relativeTimeInMillis() - startTimeMs >
+                                                        clusterStateRequest.masterNodeTimeout().millis()) {
+                                            throw new ElasticsearchTimeoutException("Timed out getting cluster state");
+                                        }
+                                        builder.startObject();
+                                        if (clusterStateRequest.waitForMetadataVersion() != null) {
+                                            builder.field(Fields.WAIT_FOR_TIMED_OUT, response.isWaitForTimedOut());
+                                        }
+                                        builder.field(Fields.CLUSTER_NAME, response.getClusterName().value());
+                                        ToXContent.Params params = new ToXContent.DelegatingMapParams(
+                                                singletonMap(Metadata.CONTEXT_MODE_PARAM, Metadata.CONTEXT_MODE_API), request);
+                                        response.getState().toXContent(builder, params);
+                                        builder.endObject();
+                                        return new BytesRestResponse(RestStatus.OK, builder);
+                                    }
+                                }.onResponse(response)));
```

Comment:
Seems a bit weird to create the listener just to immediately complete it. Why not inline this?

Reply:
Did the same for the mappings, so I didn't have to duplicate the logic in `RestBuilderListener` but now that we have two spots that follow this pattern I think we can refactor this in a follow-up and extract the logic for writing out the response somewhere so we don't have to do this.